File: src/DesignPatterns/src/Java/002-GOF-Patterns/003-BehavioralPatterns/002-Iterator/IntRange/src/org/csystem/util/IntRange.java
package org.csystem.util;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class IntRange implements Iterable<Integer> {
    private final int m_min;
    private final int m_max;
    private final int m_inc;

    public IntRange(int min, int max)
    {
        this(min, max, 1);
    }

    public IntRange(int min, int max, int inc)
    {
        m_min = min;
        m_max = max;
        m_inc = inc;
    }

    @Override
    public Iterator<Integer> iterator()
    {
        return new Iterator<Integer>() {
            private int m_index = -1;

            @Override
            public boolean hasNext()
            {
                return m_min + (m_index + 1) * m_inc <= m_max;
            }

            @Override
            public Integer next()
            {
                if (!hasNext())
                    throw new NoSuchElementException("No such element");

                return m_min + ++m_index * m_inc;
            }
        };
    }
}

File: src/DesignPatterns/src/C++/002-GOFPatterns/003-BehavioralPatterns/002-Iterator/SimpleRangeClass/IntRange.cpp
#include <stdexcept>
#include "IntRange.h"

using namespace std;

IntRange::IntRange(int min, int max, int inc)
{
	if (min > max)
		throw out_of_range("min > max");

	if (inc <= 0)
		throw out_of_range("inc < 0");

	m_min = min;
	m_max = max;
	m_inc = inc <= 0 ? 1 : inc;
}

IntRange::iterator &IntRange::iterator::operator++()
{
	//...
	m_value += m_inc;

	return *this;
}

int IntRange::iterator::operator *() const
{
	return m_value;
}

bool IntRange::iterator::operator ==(const IntRange::iterator &r) const
{
	return m_value >= r.m_value;
}

bool IntRange::iterator::operator !=(const IntRange::iterator &r) const
{
	return !(*this == r);
}